Rework multiapp iteration coupling to allow different fixed point iteration methods

GiudGiud opened the pull request:
Refs #17479 
Adds two new methods for coupling multiapps together: Secant and Steffensen. Their convergence rates are super-linear and quadratic respectively, so may beat Picard at tight coupling. 
Makes code more modular in that section

A number of better variable names thanks to @loganharbour
Full support of previous Picard parameters using deprecation (but _picard_solve attribute is removed, which is why Griffin isnt building)

Adds the possibility to relax/accelerate postprocessors on top of variables. The reason for that is that a Secant method or any Newton method is unlikely to be well-behaved over the entire domain. So variable relaxing may not work. It s also quite common to couple multiapps using postprocessor transfers, such as thermal hydraulics loops. 

Making the PR early for comments by @rwcarlsen and @YaqiWang , I still need to check that everything works as expected in corner cases and add tests. 

- [x] Update old test syntax
- [x] Add subapp tests for new methods
- [x] Add postprocessor coupling tests for all methods
- [x] Add subapp postprocessor coupling tests for new methods
- [x] Add documentation to new methods

GiudGiud committed:
Rework multiapp iteration coupling to allow multiple methods
Add minimum number of Picard iterations
Add secant solve and steffensen solve
Update tests to use new parameters
Refs #17479

GiudGiud committed:
Remove deprecated picard syntax
Consolidate catch up test folders for picard
Add stubs for postprocessor coupling tests

rwcarlsen reviewed (commented):
Upon first skim - I think this looks fairly reasonable.  You managed to avoid really changing the execution flow and just generalized the existing classes/structure a bit.  I'll take some more time to look at this in a day or two.

GiudGiud committed:
Remove NumPicardIts from tests

friedmud commented:
My first comment here is that the state saving should be done via backup/restore.  Anything else and you won't actually capture the real state (for instance - you are currently not saving stateful material properties).

YaqiWang commented:
I'd like to see more documentations on the design and the methods. I do not quite like some parameter names like `coupling_max_its` because the solve object is more general in sense that the multiapp is not necessarily another physics or even the multiapp may not exist at all. `iterative_max_its` could be better. I will need more time on reviewing it.

GiudGiud committed:
Add postprocessor coupling tests.
Add new app coupling methods tests.
Rework Secant to act on every time step.
Clean up old postprocessor storage.
Use better names for previous values of variables.
Add documentation for new app couplings.
Add test for changeOverPicard postprocessor.

GiudGiud commented:
@friedmud I have not changed the state saving. 
iterativeMultiAppCoupling calls  
`` _problem.backupMultiApps(EXEC_TIMESTEP_BEGIN);
  _problem.backupMultiApps(EXEC_TIMESTEP_END);``
and restore later on.

@YaqiWang I added a page of documentation under syntax/Executioner.
I see your point. I don't see an issue with renaming all the 'coupling' variables. I think `iterative` can replace `coupling` almost everywhere. I'll do it with the first review.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/725247/) on 631c21c wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/17593/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/17593/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 0c08243d3ae0bd6d64527bb2351f4a6c937d7c49`

moosebuild commented on libmesh:
```diff
@@ -1 +1 @@
-Subproject commit 27141d18f3137f77e33cdb3d565fd38ebfbfc46f
+Subproject commit ed76100f41840c34c146dd0a7f1ff5410370b489
```
**Caution!** This contains a submodule update

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/725306/) on aa372f7 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/17593/site)

This comment will be updated on new commits.

GiudGiud committed:
Change from IterativeMultiapp to FixedPoint
Update test inputs
Make more members private
Make more member functions protected

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/729126/) on bbf2d4c wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/17593/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/17593/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 0c08243d3ae0bd6d64527bb2351f4a6c937d7c49`

rwcarlsen commented on framework/include/executioners/FixedPointSolve.h:
```diff
@@ -0,0 +1,254 @@
+  virtual void savePreviousPostprocessorValuesAsSubApp() = 0;
+
+  /// Whether to use the fixed point algorithm (relaxed Picard, Secant, ...) instead of Picard
+  virtual bool useFixedPointAlgorithmUpdate(bool as_main_app) = 0;
```
Isn't picard itself a fixed point algorithm though (you say use picard instead of picard here :P ?  I think this needs a different/better name and doc comment.  Like solutionUpdateReady - or something along those lines.

GiudGiud replied:
Picard is the default algorithm. Relaxed Picard isnt quite Picard. 
updating the postprocessors isnt really updating the solution though. It's more often updating the boundary conditions. 
Also 'ready' isnt that clear for Steffensen where you purposefully skip a step

rwcarlsen replied:
Yeah - my suggestion wasn't really a good one either.  Maybe `fallbackToPicard` might be better...  The doc comment could certainly be better as well.  All these (pure) virtual functions could probably use some good doc comments and names so when people come along later to modify or implement new algos, they won't just be blindly copy-pasting - or tracking you down forcing you to remember why you did something a certain way :P .

rwcarlsen commented on framework/src/executioners/SecantSolve.C:
```diff
@@ -0,0 +1,271 @@
+}
+
+void
+SecantSolve::transformPostprocessorsAsMainApp()
+{
+  // Relax postprocessors for the main application
+  for (size_t i = 0; i < _transformed_pps.size(); i++)
+  {
+    // Get new postprocessor value
+    const Real current_value = getPostprocessorValueByName(_transformed_pps[i]);
+    const Real xn_m1 = _transformed_pps_values[i][1];
+    const Real fxn_m2 = _transformed_pps_values[i][2];
+    const Real xn_m2 = _transformed_pps_values[i][3];
+
+    // Save fxn_m1, received or computed before the solve
+    _transformed_pps_values[i][0] = current_value;
+
+    // Compute and set relaxed value
+    Real new_value = current_value;
+    if (!MooseUtils::absoluteFuzzyEqual(current_value - xn_m1 - fxn_m2 + xn_m2, 0))
+      new_value = xn_m1 - (current_value - xn_m1) * (xn_m1 - xn_m2) /
+                              (current_value - xn_m1 - fxn_m2 + xn_m2);
+
+    // Relax update if desired
+    new_value = _relax_factor * new_value + (1 - _relax_factor) * xn_m1;
+
+    _problem.setPostprocessorValueByName(_transformed_pps[i], new_value);
+  }
+}
+
+void
+SecantSolve::transformPostprocessorsAsSubApp()
+{
+  // Update the postprocessors
+  for (size_t i = 0; i < _secondary_transformed_pps.size(); i++)
+  {
```
These two functions seem to share a lot of implementation.  Also all the allocateStorage... savePrevious... transform... functions really seem to have a lot of redundant feeling to them - like boiler plate that doesn't quite feel right design-wise.  Also the transform... functions really are the "main" functions of these objects/algorithms - so it feels like these should be named like computeSolutionUpdate or something more evoking of their importance.

I'm not sure what the right solution is, but this part of your API design doesn't feel quite right.  Maybe you could try to explain some of the constraints/issues you ran into that lead you to designing it this way?  The asSubApp and asMainApp diversion is a bit odd - if feels like the algorithm shouldn't care about that - maybe we pass in the primary vs secondary vectors/pps as arguments and the algorithm just blindly operates on what it is given instead of having the two separate functions.  That may not really make sense/work, I'm just spitballing here...  Mostly the main vs sub app bifurcation just seems rough and requires a lot of manual "doing the right thing" - instead of allowing the code to do things the right way for us.

GiudGiud replied:
That's a good idea. I could probably consolidate the transform, maybe the save, functions. We'll loose some explicitness as to what is being updated/transformed but it'll be easier to maintain each method. 
Part of the reason why these functions were separated was how they came about, and that's no longer a constraint in the current shape of the code.

I like transformed more than updated. Like the action of the method is to transform the sequence.

rwcarlsen reviewed (dismissed):
It's hard for me to see how much of FixedPointSolve is just copied from the old PicardSolve vs how much is new/modified.  Could you comment a bit on what parts are new/different so I can focus a bit more on reading through mods?

GiudGiud commented on framework/include/postprocessors/ChangeOverPicardPostprocessor.h:
```diff
@@ -0,0 +1,55 @@
+
+#include "GeneralPostprocessor.h"
+
+class ChangeOverPicardPostprocessor;
```
yes, this is why doc build is still failing

GiudGiud commented on framework/src/executioners/Executioner.C:
```diff
@@ -90,6 +96,14 @@ Executioner::Executioner(const InputParameters & parameters)
   _fe_problem.numGridSteps(_num_grid_steps);
+
+  // Instantiate the SolveObject for the fixed point iteration algorithm
+  if (_iteration_method == "picard")
+    _fixed_point_solve = std::shared_ptr<FixedPointSolve>(std::make_shared<PicardSolve>(this));
+  else if (_iteration_method == "secant")
+    _fixed_point_solve = std::shared_ptr<FixedPointSolve>(std::make_shared<SecantSolve>(this));
+  else if (_iteration_method == "steffensen")
+    _fixed_point_solve = std::shared_ptr<FixedPointSolve>(std::make_shared<SteffensenSolve>(this));
```
That would require a subblock in the executioner for the fixed point algorithm.
It would break every app making use of Picard pretty much.

GiudGiud replied:
I ll try to see if I can make the default not use the new syntax

GiudGiud replied:
ok it turns out to be kind of hacky. SolveObjects were not really thought that way.
It would also require its own category, which might be a little premature right now.

fdkong replied:
It is not a bad idea to have sub block. Meanwhile, I also expect the code should without a sub block

GiudGiud commented on framework/include/executioners/FixedPointSolve.h:
```diff
@@ -0,0 +1,254 @@
+
+protected:
+  /// Save the variable values as a SubApp
+  virtual void savePreviousVariableValuesAsSubApp() = 0;
```
I will add more doc comment.

GiudGiud commented on framework/include/executioners/FixedPointSolve.h:
```diff
@@ -0,0 +1,254 @@
+  virtual void savePreviousVariableValuesAsSubApp() = 0;
+
+  /// Save the postprocessor values as a SubApp
+  virtual void savePreviousPostprocessorValuesAsSubApp() = 0;
```
name is as descriptive as it can be tbh

rwcarlsen replied:
Then have the doc comment describe how/when the function should be/is used.

GiudGiud commented:
I moved large chunks of the old PicardSolve into functions. So the print, save and transform(Variables, pp is new) are just cut off from the old gigantic solve routines.

GiudGiud committed:
Address Robert's review:
- use single implementation of functions for subapp/mainapp updating
- add docstrings to function definitions

Fix some test documentation
Properly rename all ChangeOverPicard
Modify tests so that they avoid issue #17709 when doing recover

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/730699/) on 209647e wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/17593/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/17593/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 0c08243d3ae0bd6d64527bb2351f4a6c937d7c49`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/731668/) on b409993 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/17593/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/17593/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 0c08243d3ae0bd6d64527bb2351f4a6c937d7c49`

GiudGiud committed:
Rebase on next:
- array variable support, #17659
- more decimals on postprocessor output, #17890
Add an assert for when there are no variables specified in the AllLocalDofsIndicesThread loop
Double cover it in FixedPoint by avoiding using this object when no variable is transformed

Fix recover tests for Picard postprocessor coupling

Move postprocessor saving after problem timestep_begin execution,
to support postprocessor relaxation which would depend on it

Address Yaqi's review:
- use tags for new vectors
- clean up master references
- have 1 min fixed point iteration
- add a word about Transient exec in auto_advance
- remove legacy param for SecantSolve
- fix docstring of FixedPointSolve::SolveStep
- improve documentation

Refs #17479

fdkong commented on framework/src/executioners/FixedPointSolve.C:
```diff
@@ -0,0 +1,557 @@
+  // multiapp solve
+  _problem.backupMultiApps(EXEC_TIMESTEP_BEGIN);
+  _problem.backupMultiApps(EXEC_TIMESTEP_END);
+
```
Do we really need to a double backup here ?

GiudGiud replied:
I did not change that part.

fdkong commented on framework/src/executioners/FixedPointSolve.C:
```diff
@@ -0,0 +1,557 @@
+      {
+        // For every iteration other than the first, we need to restore the state of the MultiApps
+        _problem.restoreMultiApps(EXEC_TIMESTEP_BEGIN);
+        _problem.restoreMultiApps(EXEC_TIMESTEP_END);
```
Hmmm... double

fdkong commented on test/tests/multiapps/secant_postprocessor/transient_sub.i:
```diff
@@ -0,0 +1,74 @@
+  petsc_options_value = 'hypre boomeramg'
+  nl_abs_tol = 1e-14
+
+  fixed_point_algorithm = 'secant'
```
I am curious why the sub app needs  to know what algorithm we are using on the top level

GiudGiud replied:
the subapp executioner is taking care of relaxing the variables specified in the main app Multiapp block

GiudGiud replied:
I agree it's not ideal.

fdkong commented on test/tests/multiapps/steffensen/transient_main.i:
```diff
@@ -0,0 +1,110 @@
+
+  fixed_point_algorithm = 'steffensen'
+  fixed_point_max_its = 30
+  transformed_variables = 'u'
```
Should we autodetect this from the transfer block?

GiudGiud replied:
u does not need to be transferred. Imagine a pp coupling with relaxation of the variables

fdkong reviewed (changes requested):
1) We rename PicardSolve to RegularPicardSolve or something. Otherwise, it hard to focus on the changes you made in `fixedPointIterration` 

2) Why sub apps need to know what algorithms ("secant") we select for the outer loop?

GiudGiud commented on framework/include/reporters/IterationInfo.h:
```diff
@@ -25,15 +25,14 @@ class IterationInfo : public GeneralReporter
   unsigned int & _num_linear;
   unsigned int & _num_nonlinear;
-  unsigned int & _num_picard;
+  unsigned int & _num_coupling;
```
I ll change that to num_fixed_point
I initially thought that was the only use

GiudGiud commented:
But regularPicardSolve is not relaxed Picard. It's just Picard. So yeah the name might be confusing
See comment. Their executioner takes care of the update, so if you want Secant method for their variables, you need to specify it there. I agree it's not optimal. It's not a change I made

fdkong commented:
>But regularPicardSolve is not relaxed Picard. It's just Picard. So yeah the name might be confusing

Name is fine. Motivation for me right now is that we want github to show only the  changes you made.  Right now, git shows you changed everything in `FixedPointSolve`. And it is hard to do code reviews

GiudGiud commented:
I did the rename, still doesnt show the diff. 
I can temporarily rename FixedPointSolve to PicardSolve so you get the diff for the review if you want

rwcarlsen commented:
@GiudGiud - have things settled enough for me to take another look, or are you still modding stuff?

YaqiWang reviewed (changes requested):
Sorry, I added comments based on old commits. Some of them may be invalid but some of them may still. In the future, you should squash the commits before asking for reviews.

YaqiWang commented on framework/src/executioners/SecantSolve.C:
```diff
@@ -0,0 +1,190 @@
+  : IterativeMultiAppSolve(ex)
+{
+  // Store a copy of the solution before the previous solution here
+  _problem.getNonlinearSystemBase().addVector("transformed_older", false, PARALLEL);
```
With this we actually can bring in part (the coupling solution part) of the capabilities in `FixedPoint` into this solve object.

YaqiWang commented on framework/src/executioners/Executioner.C:
```diff
@@ -31,6 +31,12 @@ Executioner::validParams()
   params += Reporter::validParams();
```
You possibly want to replace `PicardSolve::validParams()` above with `FixedPointSolve::validParams()`. We assume all three methods `picard`, `secant` and `steffensen` share the same set of valid params thus we can have that `fixed_point_algorithm` parameter.

GiudGiud replied:
If I do that I think I have to handle all the deprecated Picard parameters in FixedPointSolve. 
We may not assume this in the future. Some methods could have more parameters.

GiudGiud commented on framework/src/executioners/IterativeMultiAppSolve.C:
```diff
@@ -0,0 +1,534 @@
+    _problem.getNonlinearSystemBase().addVector("transformed_old", false, PARALLEL);
+
+  if (_transformed_vars.size() > 0 && _transformed_pps.size() > 0)
+    mooseWarning("Both variable and postprocessor transformation are active. If the two share dofs, the "
```
I'd need to double check with a custom crafted example, but if you apply the transform twice (even relaxed Picard) on some dofs, you ll get a different answer than once

GiudGiud commented on test/tests/multiapps/picard_catch_up_keep_solution/tests:
```diff
@@ -1,13 +0,0 @@
-[Tests]
```
It's moved not removed. I consolidated just a little bit the picard tests, there's a lot of folders

GiudGiud commented:
@YaqiWang @rwcarlsen there has been large changes because I'm addressing your reviews. 
I ll address the new ones soon, and rename back to PicardSolve since the new name didnt help things. Things will change again then. As you prefer @rwcarlsen as to when you review again, both work for me

GiudGiud commented on framework/src/executioners/IterativeMultiAppSolve.C:
```diff
@@ -0,0 +1,534 @@
+{
+  TIME_SECTION(_coupling_timer);
+
+  Real current_dt = _problem.dt();
```
It ll be in a separate PR. 
I agree isolating each part of the process more thoroughly would help here.

GiudGiud commented on framework/src/executioners/PicardSolve.C:
```diff
@@ -9,512 +9,233 @@
-                       bool relax,
-                       const std::set<dof_id_type> & relaxed_dofs)
+void
+PicardSolve::updatePostprocessorsAsMainApp()
```
Updated now.
It needs to be separate because the postprocessors and the variables arent updated at the same time.

GiudGiud commented on framework/src/postprocessors/ChangeOverPicardPostprocessor.C:
```diff
@@ -0,0 +1,130 @@
+defineLegacyParams(ChangeOverPicardPostprocessor);
+
+InputParameters
+ChangeOverPicardPostprocessor::validParams()
```
It s useful to study if Picard iterations are even necessary / should they be bound, using csv output rather than parsing of logs

GiudGiud commented on framework/src/executioners/IterativeMultiAppSolve.C:
```diff
@@ -78,26 +80,30 @@ IterativeMultiAppSolve::validParams()
-                                                      std::vector<std::string>(),
-                                                      "Relaxed variables is deprecated, use transformed_variables instead.",
-                                                      "List of master app variables to relax during coupling iterations");
+  params.addParam<std::vector<std::string>>(
```
AllLocalDof.. object expects a string. I tried to change everyone but then eventually hit that var.name() returns a string. So I think this will be for later.

GiudGiud commented on framework/src/executioners/IterativeMultiAppSolve.C:
```diff
@@ -195,9 +190,9 @@ IterativeMultiAppSolve::solve()
+  // Prepare to relax variables as a main app
   std::set<dof_id_type> transformed_dofs;
-  if (_relax_factor != 1.0)
+  if (_relax_factor != 1.0 || !dynamic_cast<PicardSolve *>(this))
```
It s one more layer of abstraction in a way. If we are going to do more than save and transform I m all for it. If not it makes it more complicated

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/748261/) on c7d56c8 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/17593/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/17593/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 27229dd4325d8f0fe75714630404ba103eaaabef`

GiudGiud commented:
@YaqiWang could you please check the use of tagging in SecantSolve.C ? I m new to this system. If it s correct I ll expand it to PicardSolve and SteffensenSolve

All your comments should be addressed / replied to otherwise.
With regards to using fixed point algorithms on postprocessors over variables, it has two distinct advantages:
- since it s only a couple values instead of a field of values, it s less likely to crash with a bad denominator in the Secant / Steffensen updates
- for the same reason, it ll be very cheap and more stable to form the Shanks / Aitken transform determinants if we want to add these methods. These should be the best in that class of methods.

YaqiWang commented on framework/src/executioners/SecantSolve.C:
```diff
@@ -0,0 +1,268 @@
+  // TODO: We would only need to store the solution for the degrees of freedom that
+  // will be transformed, not the entire solution.
+  // Store solution vectors for the two previous points and their evaluation
+  auto tag_id = _problem.addVectorTag(xn_m1_name, Moose::VECTOR_TAG_SOLUTION);
```
The usage is correct. You may want to save off this tag ids, so that you can use them in `saveVariableValues` function. In this way, you can also avoid the duplicated hard-coded solution vector names.

YaqiWang commented on test/tests/multiapps/picard_multilevel/picard_master.i:
```diff
@@ -62,12 +62,11 @@
-  picard_rel_tol = 1e-8
-  picard_abs_tol = 1e-9
-#  picard_force_norms = true
+  fixed_point_rel_tol = 1e-8
```
My big request is that you need to patch apps before this can be merged. That can be painful.

YaqiWang replied:
This touches the key capability of Griffin and we do not want to see lots of deprecation messages.

GiudGiud replied:
Yup I'll do that

fdkong replied:
deprecation is actually good idea for this use. Otherwise, it will break many input files that use `picard_max_its` and other old parameters.

We need to give users time to adjust these API changes

YaqiWang commented on framework/include/executioners/FixedPointSolve.h:
```diff
@@ -0,0 +1,287 @@
+template <>
+InputParameters validParams<FixedPointSolve>();
+
+class FixedPointSolve : public SolveObject
```
Note that this name is very similar as the one in test `FixedPoint`.

GiudGiud replied:
I think it's fine for test objects to have close names. We are not encouraging their use

YaqiWang commented on framework/doc/content/syntax/Executioner/FixedPointAlgorithms/index.md:
```diff
@@ -0,0 +1,61 @@
+Because MultiApp allows wrapping another levels of MultiApps, the design enables multi-level app coupling iterations automatically.
+
+Regardless of the fixed point algorithm used, solution vectors can be relaxed to improve the stability of the convergence.
+When a MultiApp is a sub-app and the main application of its own sub-apps, MOOSE allows relaxation of the MultiApp solution
```
This sentence is quit hard to understand, maybe rephrase a little bit?

GiudGiud replied:
changed

YaqiWang commented on framework/doc/content/syntax/Executioner/FixedPointAlgorithms/index.md:
```diff
@@ -0,0 +1,61 @@
+problem before computing the next term, this method is expected to be slower than the secant method. A poor initial guesses can also prevent convergence.
+
+!alert note
+When using the secant or Steffensen's methods, only specify variables and postprocessors from either the main application or the sub-applications to be accelerated. Specifying variables or postprocessors to be transformed in both applications will not provide as much acceleration, due to the current implementation of the methods. Future work may remove this limitation.
```
transformed?

GiudGiud replied:
transformed is used inside the code. Let s just say accelerated here

YaqiWang commented on framework/doc/content/syntax/Executioner/FixedPointAlgorithms/index.md:
```diff
@@ -0,0 +1,61 @@
+# Fixed point iteration algorithms
+
+MOOSE provides fixed point algorithms in all its executioners. This can be used to iterate a single
+application solve to converge a parameter, for example converge the mass flow rate of a fluid simulation with a target pressure drop.
```
I suggest to add some words in here to clarify the difference between this solve object and `FixedPoint` object like: it is noted that the lagged values can be stored in auxiliary variables, postprocessor values and/or user objects that are updated on `TIMESTEP_BEGIN` or `TIMESTEP_END` but we do not attempt to lag the solution vector for primal variables in this algorithm.

YaqiWang replied:
I actually want these two solve objects to be merged. But this requires the vector adding all the time, which could be a waste if no lagged solution variables are used. So I am ok if we do not merge the two. `FixedPoint` is quite simple and provides a very useful feature actually.

GiudGiud replied:
FixedPoint is a test object. I dont think we should refer to it in the documentation, that could get confusing.

I agree it's useful, but then it should be moved to the framework in a separate PR

YaqiWang replied:
I was not asking for adding `FixedPoint` in that sentence. Please add the sentence because I think it is important to clarify this.

GiudGiud replied:
ok added a note lower.

loganharbour commented on framework/include/postprocessors/ChangeOverFixedPointPostprocessor.h:
```diff
@@ -0,0 +1,55 @@
+#pragma once
+
+#include "GeneralPostprocessor.h"
+
+class ChangeOverFixedPointPostprocessor;
+
+template <>
+InputParameters validParams<ChangeOverFixedPointPostprocessor>();
```
```suggestion
```

loganharbour commented on framework/src/multiapps/MultiApp.C:
```diff
@@ -778,16 +789,18 @@ MultiApp::createApp(unsigned int i, Real start_time)
-  }
+  // Transfer coupling relaxation information to the subapps
+  auto fixed_point_solve = _apps[i]->getExecutioner()->fixedPointSolve();
+  fixed_point_solve->setMultiAppRelaxationFactor(getParam<Real>("relaxation_factor"));
+  fixed_point_solve->setMultiAppTransformedVariables(
+      getParam<std::vector<std::string>>("transformed_variables"));
+  // Handle deprecated parameter
+  if (!parameters().isParamSetByAddParam("relaxed_variables"))
+    fixed_point_solve->setMultiAppTransformedVariables(
+        getParam<std::vector<std::string>>("relaxed_variables"));
+  fixed_point_solve->setMultiAppTransformedPostprocessors(
+      getParam<std::vector<PostprocessorName>>("transformed_postprocessors"));
+  fixed_point_solve->allocateStorage(false);
```
```suggestion
  fixed_point_solve.setMultiAppRelaxationFactor(getParam<Real>("relaxation_factor"));
  fixed_point_solve.setMultiAppTransformedVariables(
      getParam<std::vector<std::string>>("transformed_variables"));
  // Handle deprecated parameter
  if (!parameters().isParamSetByAddParam("relaxed_variables"))
    fixed_point_solve.setMultiAppTransformedVariables(
        getParam<std::vector<std::string>>("relaxed_variables"));
  fixed_point_solve.setMultiAppTransformedPostprocessors(
      getParam<std::vector<PostprocessorName>>("transformed_postprocessors"));
  fixed_point_solve.allocateStorage(false);
```

loganharbour commented on framework/src/executioners/FixedPointSolve.C:
```diff
@@ -0,0 +1,561 @@
+{
+  InputParameters params = emptyInputParameters();
+
+  params.addParam<unsigned int>(
```
Many of these params should probably be range checked

loganharbour reviewed (changes requested):
I'm  through the header files and a little of the source. not done

GiudGiud commented on framework/include/executioners/SteffensenSolve.h:
```diff
@@ -0,0 +1,81 @@
+#pragma once
+
+#include "FixedPointSolve.h"
+#include "NonlinearSystem.h"
```
it's for the non unity build I think

GiudGiud commented on framework/include/executioners/FixedPointSolve.h:
```diff
@@ -0,0 +1,287 @@
+class FixedPointSolve : public SolveObject
+{
+public:
+  FixedPointSolve(Executioner * ex);
```
doubt it. But yeah the whole thing is written with pointers, I was not trying to change that

loganharbour replied:
if you don't, I will - it's a few minutes of time.

GiudGiud replied:
no worries it s on the list now

GiudGiud committed:
Remove some legacy params
Add more range checking to Fixed Point solve inputs
Use a reference for the executioner, and break every app using solve objects

Co-authored-by: Logan Harbour <loganhharbour@gmail.com>

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/748865/) on 1fc53c3 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/17593/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/17593/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 27229dd4325d8f0fe75714630404ba103eaaabef`

GiudGiud commented on framework/include/executioners/Executioner.h:
```diff
@@ -113,11 +116,29 @@ class Executioner : public MooseObject,
+  }
+
+  /// Return a pointer to the underlying fixed point solve object.
+  std::shared_ptr<FixedPointSolve> fixedPointSolve() { return _fixed_point_solve; }
```
was mimicking PicardSolve() to avoid breaking griffin. But I guess I ll be PRing Griffin anyway now

GiudGiud committed:
Add note on lagging variables in fixed point
Add more ranged param checking
Use the right tag for PicardSolve secondary relaxed

Refs #17479

GiudGiud commented:
@rwcarlsen @loganharbour @YaqiWang should be good for more review.

@YaqiWang any idea why this is breaking griffin? https://gitlab.software.inl.gov/idaholab/griffin/-/merge_requests/244/pipelines
seems like a lot of NDA tests are affected. I did the trivial name replacements, must be something a little more tricky.

YaqiWang commented:
Let me take a look.

YaqiWang commented:
It is a build failure. @GiudGiud I guess that is not what you want me to check?

GiudGiud commented:
ah I forgot about that.
Yeah it does not fail if you use the moose submodule from my branch, only with the current submodule.

YaqiWang commented:
Need you give me access to your Griffin fork.

GiudGiud commented:
Just added you

YaqiWang commented:
@GiudGiud before I look it, can you do some `grep` for all the parameters in Griffin source? Griffin uses actions to change the default values of some parameters, which could be the cause of the failure.

YaqiWang commented:
@GiudGiud Thanks for the fixes in Griffin. You did great! The only thing missing is changing `augmentedPicardConvergenceCheck` to `augmentedFixedPointConvergenceCheck` in four files. I did not check why we need to do that although it seems we do not have to with deprecation. There are two more tests failures I am currently looking, not sure if those are related at this moment.

GiudGiud commented:
Ah good catch about the augmentedPicardConvergenceCheck. I forgot about that routine. Thanks!

fdkong commented on framework/include/executioners/PicardSolve.h:
```diff
@@ -9,7 +9,8 @@
 
-#include "SolveObject.h"
+#include "FixedPointSolve.h"
+#include "NonlinearSystem.h"
```
We should use `NonlinearSystemBase.h`. I would assume Picard should work for `nonlineareigensystem` as well

GiudGiud replied:
Ok making that change. Though there are definitely better methods than secant and steffensen to solve eigenvalue problems.

fdkong replied:
Thanks. Yes, just for consistency ,,

fdkong commented on framework/include/executioners/PicardSolve.h:
```diff
@@ -19,166 +20,76 @@ class PicardSolve;
+   * @param primary Whether this routine is used for the primary transformed
+   *                quantities (as main app) or the secondary ones (as a subapp)
+   */
+  virtual bool useFixedPointAlgorithmUpdateInsteadOfPicard(const bool primary) override final;
```
I feel like the update algorithm should be implemented  in each individual  subclass. Then we could avoid this function

fdkong commented on framework/include/executioners/PicardSolve.h:
```diff
@@ -19,166 +20,76 @@ class PicardSolve;
+  TagID _old_tag_id;
+
+  /// Vector tag id for the previous solution variable, as a sub app
+  TagID _secondary_old_tag_id;
```
Should we change "secondary" to "sub"?

GiudGiud replied:
meh. should we?

fdkong replied:
When I look at secondary/primary, I always think about it is a contact problem

But I do not feel strongly about it

fdkong commented on test/tests/multiapps/picard_postprocessor/steady_main.i:
```diff
@@ -0,0 +1,103 @@
+    execute_on = 'timestep_begin'
+
+    relaxation_factor = 0.8
+    transformed_postprocessors = 'from_main'
```
Could we figure out this from the transfer block automatically?

GiudGiud replied:
It does not have to be a transferred pp. It could be a pp based on a variable received

fdkong replied:
Good to know

fdkong commented on framework/include/executioners/SecantSolve.h:
```diff
@@ -0,0 +1,105 @@
+  TagID _secondary_fxn_m2_tagid;
+
+  /// Vector tag id for the solution variable two primary solves ago, as a sub app
+  TagID _secondary_xn_m2_tagid;
```
In theory, do we indeed need so many vectors when using SecantSolve?

GiudGiud replied:
we need two "points" that represent (x, f(x) - x) in two locations. So I dont know how to do it with less than 4 known values.

fdkong reviewed (commented):
Looks good

YaqiWang commented:
Possibly it is only Griffin using this function `augmentedPicardConvergenceCheck`, so we do not have to deprecate it and just remove it completely.

GiudGiud committed:
Address Fande's review, allow eigen non linear systems, refs #17479
Make not about lagging a little clearer

GiudGiud commented:
@fdkong @rwcarlsen @YaqiWang @loganharbour anything else you'd like me to change?

YaqiWang commented:
Looks good to me. I did not see myself in the list of changes requested. Hopefully this can be merged soon.

fdkong commented:
OK, I will take a last sweep. Does the test failings need to be fixed?

GiudGiud commented:
There's already a PR in griffin to pass those tests. I think we need to coordinate MOOSE and Griffin to get both merged.

fdkong reviewed (changes requested):
Overall it is good to me now.

I was wondering whether or not we should deprecate the old parameters instead of removing them right away?

fdkong commented:
> Overall it is good to me now.
> 
> I was wondering whether or not we should deprecate the old parameters instead of removing them right away?

Never mind. We already did that

fdkong commented:
@loganharbour @rwcarlsen Could you check if your comments have been addressed? 

I will merge the PR if I do not hear anything by tomorrow noon

loganharbour commented:
> I will merge the PR if I do not hear anything back by tomorrow noon

I would suggest not dismissing reviews due to SQA :) looking this evening.

loganharbour commented on framework/doc/content/syntax/Executioner/FixedPointAlgorithms/index.md:
```diff
@@ -0,0 +1,69 @@
+## Picard fixed point iterations
+
+Picard iterations are the default fixed point iteration algorithm. They may be relaxed, with a relaxation factor specified for the
+,ain application in the `Executioner` block, and a relaxation factor specified for each `MultiApp` in their respective block.
```
```suggestion
primary application in the `Executioner` block, and a relaxation factor specified for each `MultiApp` in their respective block.
```

loganharbour commented on framework/doc/content/syntax/Executioner/FixedPointAlgorithms/index.md:
```diff
@@ -0,0 +1,69 @@
+
+## Secant method
+
+The secant method is a root finding technique which follows secant lines to find the roots of f. It is adapted here for fixed point iterations.
```
```suggestion
The secant method is a root finding technique which follows secant lines to find the roots of $f$. It is adapted here for fixed point iterations.
```

loganharbour commented on framework/doc/content/syntax/Executioner/FixedPointAlgorithms/index.md:
```diff
@@ -0,0 +1,69 @@
+,ain application in the `Executioner` block, and a relaxation factor specified for each `MultiApp` in their respective block.
+
+Relaxed Picard fixed point iterations may be described by:
+\begin{equation}
+x_{n+1} = \alpha f(x_n) + (1 - \alpha) x_n
+\end{equation}
```
```suggestion
!equation
x_{n+1} = \alpha f(x_n) + (1 - \alpha) x_n
```

loganharbour commented on framework/doc/content/syntax/Executioner/FixedPointAlgorithms/index.md:
```diff
@@ -0,0 +1,69 @@
+
+The secant method is a root finding technique which follows secant lines to find the roots of f. It is adapted here for fixed point iterations.
+The secant method may be described by:
+\begin{eqnarray}
+x_{n+1} = x_n - \dfrac{(f(x_n) - x_n) * (x_n - x_{n-1})}{f(x_n) - x_n - f(x_{n-1}) + x_{n-1}}
+\end{eqnarray}
```
```suggestion
!equation
x_{n+1} = x_n - \dfrac{(f(x_n) - x_n) * (x_n - x_{n-1})}{f(x_n) - x_n - f(x_{n-1}) + x_{n-1}}
```

loganharbour replied:
This doesn't even format as shown right now

loganharbour commented on framework/doc/content/syntax/Executioner/FixedPointAlgorithms/index.md:
```diff
@@ -0,0 +1,69 @@
+with the same conventions as above. The relaxation factor, if used, is not shown here. The secant method is easily understood for 1D problems,
+where $(x_n, f(x_n) - x_n)$ are the coordinates of the points used to draw the secant, of slope $\dfrac{x_n - x_{n-1}}{(f(x_n) - x_n) - (f(x_{n-1}) - x_{n-1})}$.
+
+Convergence of the secant method is expected to be super-linear when it converges, with an order of $\dfrac{1 + \sqrt{5}}{2}. Some conditions
```
```suggestion
Convergence of the secant method is expected to be super-linear when it converges, with an order of $\dfrac{1 + \sqrt{5}}{2}4. Some conditions
```

loganharbour commented on framework/doc/content/syntax/Executioner/FixedPointAlgorithms/index.md:
```diff
@@ -0,0 +1,69 @@
+
+## Steffensen's method
+
+Steffensen's method is a root finding technique based on perturbating a solution at a given point to approximate the local derivative. The update is
+then similar to Newton's method which uses the exact derivative.
+
+\begin{eqnarray}
+g(x_{n}) = \dfrac{f(x_{n} + f(x_{n}))}{f(x_{n})}
+x_{n+1} = x_n - \dfrac{f(x_n)}{g(x_{n})}
+\end{eqnarray}
+
+Convergence of Steffensen's method is expected to be quadratic when it converges. However because it requires two evaluations of the coupled
+problem before computing the next term, this method is expected to be slower than the secant method. A poor initial guesses can also prevent convergence.
```
```suggestion
Steffensen's method is a root finding technique based on perturbating a solution at a given point to approximate the local derivative, as:

!equation
g(x_{n}) = \dfrac{f(x_{n} + f(x_{n}))}{f(x_{n})}\,,

!equation
x_{n+1} = x_n - \dfrac{f(x_n)}{g(x_{n})}\,.

The update is then similar to Newton's method which uses the exact derivative.
```

or something of the sort

loganharbour commented on framework/include/executioners/Executioner.h:
```diff
@@ -15,7 +15,10 @@
 #include "FEProblemSolve.h"
+#include "FixedPointSolve.h"
 #include "PicardSolve.h"
+#include "SecantSolve.h"
+#include "SteffensenSolve.h"
```
```suggestion
```
Neither one of these is used

loganharbour commented on framework/src/executioners/FixedPointSolve.C:
```diff
@@ -0,0 +1,565 @@
+  saveAllValues(true);
+
+  if (_has_fixed_point_its)
+    _console << COLOR_MAGENTA << "\nMain app solve:\n" << COLOR_DEFAULT;
```
Didn't we switch to primary?

GiudGiud replied:
syntax/MultiApps/index.md still uses main and sub.

YaqiWang replied:
I thought we want to use main-application and sub-application?

GiudGiud replied:
I think so too.

loganharbour replied:
We're switching away from the sub description as the master description goes out, but obviously that's not everywhere yet

GiudGiud replied:
I thought we were (and pretty much are) getting rid of master-slave, not main-sub.

loganharbour replied:
🤷🏼 that's why I asked

GiudGiud committed:
Use \!equation to display fixed point equations in documentation, remove unnecessary headers from Executioner.h, refs #17479

loganharbour commented:
> 16 conversations must be resolved before merging.

GiudGiud commented:
I've expanded all conversations on the front page and I cant find them lol

loganharbour commented:
Is the Griffin patch ready to merge spare for the submodule update?

GiudGiud commented:
@YaqiWang said he was ready available to get it through on Thursday.

loganharbour commented:
> @YaqiWang said he was ready available to get it through on Thursday.

So is it? That doesn’t answer the question

GiudGiud commented:
This isn't held up by the Griffin patch. It's held up by the resolving conversation bug.
I ll let @YaqiWang speak for the readiness of the Griffin patch.

loganharbour commented:
> It's held up by the resolving conversation bug.

that’s not what I’m asking about

> I ll let @YaqiWang speak for the readiness of the Griffin patch

this is what I’m waiting for. It’s not held up by the conversation bug. Once he’s around to edit the commit with the new hash on next, we’re good to go

YaqiWang commented:
@GiudGiud did you do a rebase of your Griffin patch on top of devel after my MR merged? If so, yes, it is ready.

GiudGiud commented:
not sure which MR you are talking about, so I ll just rebase now

YaqiWang commented:
#250. I tagged you in that MR but you did not reply.

GiudGiud commented:
ok done

YaqiWang commented:
@loganharbour We can proceed to merge this now.

GiudGiud commented:
Ok should be fixed now. I never got an email for it, something must have failed in the notification. DId you add the tag while editing the comment? I m not sure this works, but I m curious if it s that
